substratevm/src/com.oracle.svm.hosted.jdk11/src/com/oracle/svm/hosted/ModuleLayerFeature.java

```diff
@@ -119,49 +139,183 @@ public void afterAnalysis(AfterAnalysisAccess access) {
+    private ModuleLayer synthesizeRuntimeBootLayer(ImageClassLoader cl, Set<String> reachableModules) {
         Configuration cf = synthesizeRuntimeBootLayerConfiguration(cl.modulepath(), reachableModules);
         try {
             ModuleLayer runtimeBootLayer = moduleLayerConstructor.newInstance(cf, List.of(), null);
-            patchRuntimeBootLayer(runtimeBootLayer, reachableModules);
-            // Ensure that the lazy field ModuleLayer.modules gets set
-            runtimeBootLayer.modules();
+            Map<String, Module> nameToModule = synthesizeNameToModule(runtimeBootLayer, cl.getClassLoader());
+            patchRuntimeBootLayer(runtimeBootLayer, nameToModule);
             return runtimeBootLayer;
         } catch (InstantiationException | IllegalAccessException | InvocationTargetException ex) {
             throw VMError.shouldNotReachHere("Failed to synthesize the runtime boot module layer.", ex);
         }
     }
 
-    private Configuration synthesizeRuntimeBootLayerConfiguration(List<Path> mp, Map<String, Module> reachableModules) {
+    private Configuration synthesizeRuntimeBootLayerConfiguration(List<Path> mp, Set<String> reachableModules) {
         ModuleFinder beforeFinder = new BootModuleLayerModuleFinder();
         ModuleFinder afterFinder = ModuleFinder.of(mp.toArray(Path[]::new));
-        Set<String> roots = reachableModules.keySet();
         try {
-            return Configuration.empty().resolve(beforeFinder, afterFinder, roots);
+            return Configuration.empty().resolve(beforeFinder, afterFinder, reachableModules);
         } catch (FindException | ResolutionException | SecurityException ex) {
             throw VMError.shouldNotReachHere("Failed to synthesize the runtime boot module layer configuration.", ex);
         }
     }
 
-    private void patchRuntimeBootLayer(ModuleLayer runtimeBootLayer, Map<String, Module> reachableModules) {
+    private void patchRuntimeBootLayer(ModuleLayer runtimeBootLayer, Map<String, Module> nameToModule) {
         try {
-            moduleNameToModuleField.set(runtimeBootLayer, reachableModules);
-            moduleParentsField.set(runtimeBootLayer, List.of(ModuleLayer.empty()));
+            moduleLayerNameToModuleField.set(runtimeBootLayer, nameToModule);
+            moduleLayerParentsField.set(runtimeBootLayer, List.of(ModuleLayer.empty()));
         } catch (IllegalAccessException ex) {
             throw VMError.shouldNotReachHere("Failed to patch the runtime boot module layer.", ex);
         }
+
+        // Ensure that the lazy modules field gets set
+        runtimeBootLayer.modules();
+    }
+
+    /**
+     * This method creates Module instances that will populate the runtime boot module layer of the image.
+     * This implementation is copy-pasted from {@link java.lang.Module#defineModules(Configuration, Function, ModuleLayer)}
+     * with few simplifications (removing multiple classloader support) and removal of VM state updates
```

Comment:
> removal of VM state updates

:grimacing: if `java.lang.Module#defineModules` would have been written in a way that partitions module creation from VM state update we could reuse the module creation part without this reflection-orgy.